table/block_based_table_reader.cc

```diff
@@ -1769,9 +1777,52 @@ bool BlockBasedTable::PrefixMayMatch(const Slice& internal_key,
   if (!prefix_extractor->InDomain(user_key)) {
     return true;
   }
-  assert(rep_->table_properties->prefix_extractor_name.compare(
-             prefix_extractor->Name()) == 0);
   auto prefix = prefix_extractor->Transform(user_key);
+  if (prefix_extractor_changed) {
+    // Try to reuse the bloom filter in the SST table if prefix_extractor in
+    // mutable_cf_options has changed. If range [user_key, upper_bound) all
+    // share the same prefix then we may still be able to use the bloom filter.
+    if (read_options.iterate_upper_bound != nullptr &&
+        rep_->table_prefix_extractor) {
+      if (!rep_->table_prefix_extractor->InDomain(user_key) ||
+          !rep_->table_prefix_extractor->InDomain(
+              *read_options.iterate_upper_bound)) {
+        return true;
+      }
+      Slice user_key_xform = rep_->table_prefix_extractor->Transform(user_key);
+      Slice upper_bound_xform = rep_->table_prefix_extractor->Transform(
+          *read_options.iterate_upper_bound);
+      // first check if user_key and upper_bound all share the same prefix
+      if (user_key_xform.compare(upper_bound_xform) != 0) {
+        auto& comparator = rep_->internal_comparator;
+        // second check if user_key's prefix is the immediate predecessor of
+        // upper_bound and have the same length. If so, we know for sure all
+        // keys in the range [user_key, upper_bound) share the same prefix.
+        // Also need to make sure upper_bound are full length to ensure
+        // correctness
+        if (read_options.iterate_upper_bound->size() !=
+                rep_->table_prefix_extractor->FullLength() ||
+            !comparator.IsSameLengthImmediateSuccessor(
+                user_key_xform.ToString(),
+                read_options.iterate_upper_bound->ToString())) {
+          // TODO(Zhongyi): delete debug code before merging
+          // fprintf(stdout, "BF = %s, internal_key = %s, upper_bound = %s, user_key_xform = "
+          // "%s, upper_bound_xform = %s transformed into different prefix, not "
+          // "possible\n",
+          //         rep_->table_prefix_extractor->Name(),
+          //         user_key.ToString().c_str(),
+          //         read_options.iterate_upper_bound->ToString().c_str(),
+          //         user_key_xform.ToString().c_str(),
+          //         upper_bound_xform.ToString().c_str());
+          return true;
+        }
+      }
+      // Use prefix_extractor found in the SSTtable to do the transformation.
+      prefix = user_key_xform;
```

Comment:
Can you move this whole logic to the filter? We can create a new function in the filter class for it. This allows us to plug in different range filter implementation in the future.